### PR TITLE
RFC: meson: Format all meson files with muon fmt

### PR DESCRIPTION
**Pull request by medhefgo**
$ find . -name meson.build -exec muon fmt -c muon_fmt.ini -i {} +

### CODE REVIEW

**Comment by medhefgo** [comment]
I did some browsing through this and in a lot of places it fixes inconsistencies. As per usual with formatters, some things get better, some things get worse.

Imho, I think it's worth it. But I can also simply just fix the biggest outliers if we don't want this…

**Comment by annacrombie** [comment]
If it is a problem, I can add an option to disable the files() sorting behavior.

File: catalog/meson.build [hunk 1]
```diff
@@ -21,12 +22,16 @@ foreach file : in_files
...
+        )
 endforeach
 
-meson.add_install_script('sh', '-c',
-                         'test -n "$DESTDIR" || @0@/journalctl --update-catalog'.format(rootbindir))
+meson.add_install_script(
+        'sh',
+        '-c',
+        'test -n "$DESTDIR" || @0@/journalctl --update-catalog'.format(rootbindir),
+)
```
Comment (keszybz) [inline catalog/meson.build:37]: Please no.

File: docs/var-log/meson.build [hunk 2]
```diff
@@ -1,6 +1,5 @@
 # SPDX-License-Identifier: LGPL-2.1-or-later
 
 if conf.get('HAVE_SYSV_COMPAT') == 1 and get_option('create-log-dirs')
-        install_data('README.logs',
-                     install_dir : docdir)
+        install_data('README.logs', install_dir : docdir)
```
Comment (keszybz) [inline docs/var-log/meson.build:4]: As discussed previously, please don't do this.

File: hwdb.d/meson.build [hunk 3]
```diff
@@ -43,37 +45,35 @@ if conf.get('ENABLE_HWDB') == 1
...
-                                         mkdir_p.format(sysconfdir / 'udev/hwdb.d'))
+                meson.add_install_script('sh', '-c', mkdir_p.format(sysconfdir / 'udev/hwdb.d'))
 
-                meson.add_install_script('sh', '-c',
-                                         'test -n "$DESTDIR" || @0@/systemd-hwdb update'.format(rootbindir))
+                meson.add_install_script(
+                        'sh',
+                        '-c',
+                        'test -n "$DESTDIR" || @0@/systemd-hwdb update'.format(rootbindir),
+                )
```
Comment (keszybz) [inline hwdb.d/meson.build:60]: Those very carefully formatted to use the same pattern: `sh`, `-c` on one line, the command on the other. This reformatting blows up the whitespace and reduces consistency.

File: hwdb.d/meson.build [hunk 4]
```diff
@@ -43,37 +45,35 @@ if conf.get('ENABLE_HWDB') == 1
...
                 capture : true,
                 install : true,
-                install_dir: udevhwdbdir)
+                install_dir : udevhwdbdir,
+        )
 
-        install_data(hwdb_files_notest,
-                     hwdb_files_test,
-                     install_dir : udevhwdbdir)
+        install_data(hwdb_files_notest, hwdb_files_test, install_dir : udevhwdbdir)
```
Comment (keszybz) [inline hwdb.d/meson.build:51]: Hmmmmm.

File: hwdb.d/meson.build [hunk 5]
```diff
@@ -43,37 +45,35 @@ if conf.get('ENABLE_HWDB') == 1
...
-                     args : [hwdb_files_test,
-                             auto_suspend_rules],
-                     timeout : 90)
+                test(
+                        'parse-hwdb',
+                        parse_hwdb_py,
+                        suite : 'dist-check',
+                        args : [hwdb_files_test, auto_suspend_rules],
+                        timeout : 90,
+                )
```
Comment (keszybz) [inline hwdb.d/meson.build:71]: This is for the worse…

File: hwdb.d/meson.build [hunk 6]
```diff
@@ -43,37 +45,35 @@ if conf.get('ENABLE_HWDB') == 1
...
 
-run_target(
-        'update-hwdb',
-        command : [update_hwdb_sh, meson.current_source_dir()])
+run_target('update-hwdb', command : [update_hwdb_sh, meson.current_source_dir()])
 
-run_target(
-        'update-hwdb-autosuspend',
-        command : [update_hwdb_autosuspend_sh, project_source_root])
+run_target('update-hwdb-autosuspend', command : [update_hwdb_autosuspend_sh, project_source_root])
```
Comment (keszybz) [inline hwdb.d/meson.build:79]: As as those two…

File: man/meson.build [hunk 7]
```diff
@@ -5,22 +5,33 @@ subdir('rules')
 
 want_man = get_option('man')
 want_html = get_option('html')
-xsltproc = find_program('xsltproc',
-                        required : want_man == 'true' or want_html == 'true')
+xsltproc = find_program('xsltproc', required : want_man == 'true' or want_html == 'true')
```
Comment (keszybz) [inline man/meson.build:8]: Likewise.

File: man/meson.build [hunk 8]
```diff
@@ -5,22 +5,33 @@ subdir('rules')
...
+        'ansi',
+        '--stringparam',
+        'man.authors.section.enabled',
+        '0',
+        '--stringparam',
+        'man.copyright.section.enabled',
+        '0',
+        '--stringparam',
+        'systemd.version',
+        '@0@'.format(meson.project_version()),
```
Comment (keszybz) [inline man/meson.build:31]: Absolutely no.

File: man/meson.build [hunk 9]
```diff
@@ -115,17 +130,23 @@ systemd_directives_xml = custom_target(
...
-                                    ['systemd.index',      '7', systemd_index_xml]] : []
+        command : [make_man_index_py, '@OUTPUT@'] + nonindex_xml_files,
+)
+
+foreach tuple : (
+        xsltproc.found() ? [
+                ['systemd.directives', '7', systemd_directives_xml],
+                ['systemd.index', '7', systemd_index_xml],
+        ] : []
+)
```
Comment (keszybz) [inline man/meson.build:149]: No.

File: man/meson.build [hunk 10]
```diff
@@ -168,73 +191,69 @@ foreach tuple : xsltproc.found() ? [['systemd.directives', '7', systemd_directiv
...
-        command : [echo])
-
-html = custom_target(
-        'html',
-        output : 'html',
-        depends : html_pages,
-        command : [echo])
+man = custom_target('man', output : 'man', depends : man_pages, command : [echo])
+
+html = custom_target('html', output : 'html', depends : html_pages, command : [echo])
```
Comment (keszybz) [inline man/meson.build:203]: No. No.

File: man/meson.build [hunk 11]
```diff
@@ -168,73 +191,69 @@ foreach tuple : xsltproc.found() ? [['systemd.directives', '7', systemd_directiv
...
+                command : [
+                        rsync,
+                        '-rlv',
+                        '--delete-excluded',
+                        '--include=man',
+                        '--include=*.html',
+                        '--exclude=*',
+                        '--omit-dir-times', meson.current_build_dir(),
+                        get_option('www-target'),
+                ],
```
Comment (keszybz) [inline man/meson.build:218]: Meh.

**Review by keszybz — COMMENTED** [review]
I reviewed just the first two files… Sorry, but this suffers from the usual issue with automatic reformatters: the reformatter doesn't understand logic of the expression (which is important e.g. for formatting of ternary operators) or similarity between neighbouring cases (which is important e.g. for the case of repeated file lists, some with fewer and some with more entries). The patch makes some nice cleanups (whitespace around `:`, formatting of the closing brace), which we would want, but makes many unwanted changes in other places. We would certainly want a patch which is filtered to only include the former.

(This follows the general rule: we don't merge patches which fix some cases but make other places noticeably worse.)

**Comment by keszybz** [comment]
`man/rules/meson.build` is autogenerated by a script which maintains formatting and ordering stability. It shouldn't be reformatted.

File: hwdb.d/meson.build (hunk 5 above)
Comment (medhefgo) [reply hwdb.d/meson.build:71]: I think this way is far better as it works better for very long function invocations:
```
some_var = meson.do_some_very_verbose_thing(arg,
                                            arg,
                                            arg)
```
just looks silly. And mixing different styles is too.

File: man/meson.build (hunk 9 above)
Comment (medhefgo) [reply man/meson.build:149]: This should be reworked to use dicts anyway (and have it not defined inline). This is a code smell to me.

**Comment by medhefgo** [comment]
I think you're too focused on beauty over function. It's far better to have a consistent style overall than this hodgepoge of different ways to style things all over the place.

I'll admit that not everything looks perfect, but there is far more value in being able to write some code and then let the your IDE do the cleanup for you and not having to worry about making changes to unrelated areas in files.

Also, a lot of those edge-cases are simply that: edge-cases. Add another arg and it'll break up. Remove one, and it goes on a single line. That's the nature of code-formatters. And I will argue that it's a good thing, because that removes bike-shedding over formatting.

Now @annacrombie , since you're already here I have some pointers for muon fmt to improve (in order of importance):
 - sorting files lists is a good idea and should not be optional imho (we can rework how we pick a test name). In fact, it would be nice if there was a lint for this.
 - Have an option to force all functions that are called with kwargs to be multiline (`kwargs_force_multiline` is just for the literals passed into them?). Or maybe even any function with more than `n` args.
 - Maybe also allow providing a list of functions to perfer single/multiline (unless they go over `max_line_len`).
 - Either provide a comment to mark something as a table or auto-detect those so that large "pretty" tables are nicely aligned.
 - Or at least a means to opt out of formatting for the next statement/block.
 - Also, some formatters (omnisharp for example) have this notion of keeping lines that have been broken up manually instead of forcing it single-line (and the other way around). So,
   ```
   function(arg1,arg2)
   function(
        arg1,
           arg2)
   ```
   would become this (because the calls do not go over max line length):
   ```
   function(arg1, arg2)
   function(
           arg1,
           arg2,
   )
   ```
 - I like the command line argument handling where it keeps pairs of `'--option', 'value'` on one line, but this should probably also apply to regular lists, not just the literals to specific functions.

File: man/meson.build (hunk 11 above)
Comment (medhefgo) [reply man/meson.build:218]: @annacrombie This looks wrong. The cmdline option formatting should keep args isolated if they already are (unless they end with `=`).

**Comment by medhefgo** [comment]
Also, for the very same reason I would rather we drop the current clang format settings we have (except for max line length and indentation size of 4 or 8 spaces, because the default of 2 is way too dense) and enforce that. The only reason clang-format does a terrible job in a lot of places is because we have a very weird and specific set of rules in their that trip it up.
And the nice thing is there are github actions to enforce this, so you'll never get weird styles again.

(Yes, tables and some other edge-cases. We can opt out of clang-format using comments.)

File: hwdb.d/meson.build (hunk 5 above)
Comment (keszybz) [reply hwdb.d/meson.build:71]: > And the nice thing is there are github actions to enforce this, so you'll never get weird styles again.

I find that particularly obnoxious. Incidental contributors who use a different editor and have a different version of the formatting thingymagingy will get a linebreak in a spot and the patch will be "rejected". Enforced formatting raises the bar for contributions. If wouldn't be so bad if the automatic formatting was good. But as the examples here show, quite often it's at best "meh", which means that we enforce something which isn't worth enforcing.

What I'd like to see is a formatter which helps you automatically format the code you edit. I.e.: you edit a piece of code, so you touch it anyway, so it helps you DTRT, but if want to, you can format a piece of code in a different fashion. And it stays in this way until the next person comes along and touches it.

**Pull request closed without merging**